src/test/ledger/Invariants_test.cpp

```diff
@@ -0,0 +1,183 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <test/jtx.h>
+#include <test/jtx/Env.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/app/tx/apply.h>
+#include <ripple/app/tx/impl/Transactor.h>
+#include <ripple/app/tx/impl/ApplyContext.h>
+#include <boost/algorithm/string/predicate.hpp>
+
+namespace ripple {
+
+class Invariants_test : public beast::unit_test::suite
+{
+
+    class TestSink : public beast::Journal::Sink
+    {
+    public:
+        std::stringstream strm_;
+
+        TestSink () : Sink (beast::severities::kWarning, false) {  }
+
+        void
+        write (beast::severities::Severity level,
+            std::string const& text) override
+        {
+            if (level < threshold())
+                return;
+
+            strm_ << text << std::endl;
+        }
+    };
+
+    // this is common setup/method for running a failing invariant check. The
+    // precheck function is used to manipulate the ApplyContext with view
+    // changes that will cause the check to fail.
+    void
+    doInvariantCheck( bool enabled,
+        std::function < bool (
+            test::jtx::Account const& a,
+            ApplyContext& ac)> const& precheck )
+    {
+        using namespace test::jtx;
+        Env env {*this, envconfig([&enabled](std::unique_ptr<Config> cfg) {
+            cfg->section("check_invariants")
+                .set("enabled", enabled ? "1" : "0");
+            return cfg;
+        })};
+
+        Account A1 {"A1"};
+        env.fund (XRP (1000), A1);
+        env.close();
+
+        // dummy/empty tx to setup the AccountContext
+        auto tx = STTx {ttACCOUNT_SET, [](STObject&){  } };
+        OpenView ov {*env.current()};
+        TestSink sink;
+        beast::Journal jlog {sink};
+        ApplyContext ac {
+            env.app(),
+            ov,
+            tx,
+            tesSUCCESS,
+            env.current()->fees().base,
+            tapNONE,
+            jlog
+        };
+
+        BEAST_EXPECT(precheck(A1, ac));
+
+        auto tr = ac.checkInvariants(tesSUCCESS);
+        if (enabled)
+        {
+            BEAST_EXPECT(tr == tecINVARIANT_FAILED);
+            BEAST_EXPECT(boost::starts_with(sink.strm_.str(), "Invariant failed:"));
+        }
+        else
+        {
+            BEAST_EXPECT(tr == tesSUCCESS);
+            BEAST_EXPECT(sink.strm_.str().empty());
+        }
+    }
+
+    void
+    testEnabled ()
+    {
+        using namespace test::jtx;
+        testcase ("feature enabled");
+        Env env {*this};
+        auto hasInvariants =
+            env.app().config().features.find (featureEnforceInvariants);
+        BEAST_EXPECT(hasInvariants != env.app().config().features.end());
+        auto const& section = env.app().config().section("invariants");
+        BEAST_EXPECT(get<bool> (section, "enabled", true));
+    }
```

Comment:
I would also like to see `BEAST_EXPECT(env.current()->rules().enabled(featureEnforceInvariants));`